### PR TITLE
[Refinement] refactor RetryTaskMap

### PR DESCRIPTION
**Pull request by Bodong-Yang**
## Description

This PR refactors `RetryTaskMap`, which realizes the feature of the new downloading retry strategy.

### Introduction to new downloading retry strategy

The new strategy managing all OTA files downloading as a `downloading group`, and applies retry strategy in `downloaing group` scope. For each retry round, all the downloading tasks will be executed, and the failed tasks will be the input for the next retry round.
The downloading group will keep retry, until exceeding `<DOWNLOADING_GROUP_RETRY_LIMIT>`, or interrupted by custom logic(calling `shutdown` API of the `RetryTaskMap`).

Currently, otaclient uses `RetryTaskMap` with unlimited retry, and only interrupts when keeps no data transfer longer than `<DOWNLOAD_GROUP_INACTIVE_TIMEOUT=5mins>`.

### Motivation

The previous implementation of `RetryTaskMap` is complicated and with relatively lack of code readability. This PR simplifies the implementation, resolve high code cognitive complexity in some core methods, while maintaining the same funcitonality.

### Design of new RetryTaskMap

New `RetryTaskMap` comes with two classes:
1. `_TaskMap`: the executor for each retry round,
2. `RetryTaskMap`: managing retry rounds, provides `map` API for starting mapping, and `shutdown` API to interrupt the execution.

`_TaskMap` is a threadpool-based task executor for a single retry round. it has a `dispatcher` thread to dispatching tasks into the threadpool, while exposes a generator instance(returned by calling `map` method) for upper caller to yield the done task. Done tasks is collected by each task's callback, and put in to a que, while generator instance will yield from the que.
After the all the input tasks is dispatched and finished, it will return a list of failed tasks if any, or None to indicate the success of this retry round.

`RetryTaskMap` manages retry rounds, it manages instance of `_TaskMap` for each retry round, yield from the `_TaskMap` inst for the upper caller. For each retry round, if there are failed tasks from this retry round, `RetryTaskMap` will cleanup the old `_TaskMap` instance, and then init and start new instance of `_TaskMap`, with input as previous failed tasks.
It will continue the above procedure, until all the tasks are finished successfully, or being interrupted due to exceeding retry count, or by custom interruption logic.

### Improvement

1. memory usage reduced, especially when network totally cutoff, and when retry rounds processing, the memory usage almost doesn't increase,
2. performance slightly improved, 
3. code readability improved, reduce ~70 lines of code

## Check list

- [x] test file(s) that covers the change(s) is implemented.
- [x] local test is passed. 
- [x] single vm test is passed.

## Behavior changes

Does this PR introduce behavior change(s)?

- [x] No.

## Breaking change

Does this PR introduce breaking change?
- [x] No.

### CODE REVIEW

File: otaclient/app/common.py [hunk 1]
```diff
@@ -336,264 +337,190 @@ def urljoin_ensure_base(base: str, url: str):
...
+            self._se.acquire()
+            self._total_tasks_count += 1
+
+            fut = self._executor.submit(func, item)
+            fut.add_done_callback(partial(self._done_task_cb, item))
+            self._dispatched_tasks.add(fut)
+        logger.debug(f"dispatcher done: {self._total_tasks_count=}")
+        self._dispatch_done = True
+
+    def _done_task_collecter(self) -> Generator[DoneTask, None, None]:
```
Comment (obi-t4) [inline otaclient/app/common.py:414]: collecter -> collector ?

**Review by obi-t4 — COMMENTED** [review]
I left some comments, please have a check.

File: otaclient/app/common.py [hunk 2]
```diff
@@ -336,264 +337,190 @@ def urljoin_ensure_base(base: str, url: str):
...
+        add failed to failed list.
+        """
+        self._se.release()  # always release se first
+        # NOTE: don't change dispatched_tasks if shutdown_event is set
+        if self._shutdown_event.is_set():
+            return
+
+        self._dispatched_tasks.discard(fut)
+        # check if we finish all tasks
+        _done_task_num = next(self._done_task_counter)
```
[the remark this answers is not in the document]
Comment (Bodong-Yang) [reply otaclient/app/common.py:389]: Yes, `self._done_task_counter` is an instance of `itertools.count`. When using CPython(with GIL),  `itertools.count` is pure C implemented, and it is thread-safe thanks to GIL.

[the remark this answers is not in the document]
Comment (Bodong-Yang) [reply otaclient/app/common.py:389]: And we only get value from the counter, not saving it, so the value we get reflects the calls to the counter(the done task nums) at the time the `next(counter)` is executed.

File: otaclient/app/common.py (hunk 1 above)
Comment (Bodong-Yang) [reply otaclient/app/common.py]: Fixed in [7dd1586](https://github.com/tier4/ota-client/pull/232/commits/7dd1586a6a38c54facb6d8e77e861559bb6cd4b1).

**Review by obi-t4 — APPROVED** [review]
LGTM :+1:

File: otaclient/app/common.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Bodong-Yang) [reply otaclient/app/common.py:391]: Updated in [72607d2](https://github.com/tier4/ota-client/pull/232/commits/72607d22f0e5b0ca66f40d02b0fd9e21ffca3ae0),
comments related to `itertools.count` usage are added.

```
        # NOTE: itertools.count is only thread-safe in CPython with GIL,
        #       as itertools.count is pure C implemented, calling next over
        #       it is atomic in Python level.
```

**Pull request merged**